fhir-persistence-schema/docs/DB2MultiTenancy.md

```diff
@@ -1,16 +1,225 @@
+...
+COMMIT;
+DROP TABLESPACE TS_TENANT1;
+```
+
+In this case, TS_TENANT1 is the name of the tablespace previously assigned to the tenant being dropped.
+
+Note that DB2 requires a commit after the partition is detached and before the table `<tbl_drp1>` can be dropped. The tooling must implement this process to be idempotent to handle cases where a detach or drop might fail.
+
+Dropping partitions in this way works by swizzling metadata in the catalog which requires very little (redo) logging and is therefore far more efficient than deleting actual rows of data.
+
+
+
+
+## Tenant Authentication
+
+The server connect to Db2 with a user configured for the least privileges it requires to provide the FHIR resource persistence and search functions. This user does not have any database administration privileges, unlike the FHIRADMIN user which is used to deploy the schema, allocate tenants etc.
+
+![Privilege Design](privilege_design.png)
+
+To prevent an application (or other user with access to the applications database connection information) from arbitrarily setting the session_tenant value, the user used by the IBM FHIR server to connect is granted only read access to the variable. This is why we create the session_tenant variable in the FHIR_ADMIN schema, owned by the FHIRADMIN user, not the FHIRDATA schema, which is the default schema for the user used by the server to access the database. In order to set (write) its value, the data user is granted execute privileges on a DB2 stored procedure called FHIR_ADMIN.SET_TENANT. This is the only way a non-administrative user can set the session_tenant value and access the data.
+
+The following table summarizes the properties/values involved in authenticating tenant access:
+
+| Property | Description |
+|----------|-------------|
+|TENANT_NAME | Used by applications to uniquely identify tenants |
+| MT_ID | The integer primary key of the TENANTS table, used to map the TENANT_NAME to an integer value which is used as the partitioning key and row-based access permission predicate. |
+| TENANT_KEY | A 32 byte random value created when the tenant is provisioned. This is a secret, and must be passed by the client to the database in order to obtain access to a tenant. A tenant may have multiple tenant keys to support key rotation. (deploy new key before removing old key) |
+| TENANT_SALT | A 32 byte random value used as a salt when computing the SHA-256 tenant hash |
+| TENANT_HASH | A SHA-256 hash of the TENANT_KEY and TENANT_SALT. The TENANT_KEY is never stored directly by the database. |
+
+
+The FHIR_ADMIN.SET_TENANT stored procedure takes two parameters: the tenant name and the tenant key. This acts as a user/password pair tuple. The SET_TENANT code looks up a previously stored salt value associated with the tenant name and computes a hash of the salt + tenant key. If the result matches the stored hash value, the call is considered authentic, and the session_tenant variable is set.
```

Comment:
```suggestion
The FHIR_ADMIN.SET_TENANT stored procedure takes two parameters: the tenant name and the tenant key. This acts as a authorization. The SET_TENANT code looks up a previously stored salt value associated with the tenant name and computes a hash of the salt + tenant key. If the result matches the stored hash value, the call is considered authentic, and the session_tenant variable is set.
```

I suggest minimizing the user/password reference, a user reading this may conflate it with their user-password.

Reply:
fixed